src/views/SpikeLocations/SpikeLocationsWidget.tsx

```diff
@@ -0,0 +1,310 @@
+            ctxt.textBaseline = 'middle'
+            const xOffset = offsetLabels ? 1.4 * pixelRadius2 : 0
+            electrodesWithColors.forEach(e => {
+                const pt = applyAffineTransform(affineTransform, {x: e.pixelX, y: e.pixelY})
+                ctxt.fillStyle = offsetLabels ? colors.textDark : e.textColor
+                ctxt.fillText(`${e.e.label}`, pt.x - xOffset, pt.y)
+            })
+        }
+    }, [colors, offsetLabels, pixelElectrodes, pixelRadius, selectedElectrodeIds, showLabels, affineTransform, radiusScale])
+
+    const pixelPointsByUnit = useMemo(() => {
+        const ret: {[key: string | number]: {x: number, y: number}[]} = {}
+        for (let unit of units) {
+            const points = unit.xLocations.map((xLocation, i) => {
+                const x = unit.xLocations[i]
+                const y = unit.yLocations[i]
+                return transformPoint(transform, [x, y])
+            })
+            const pts = points.map(point => (applyAffineTransform(affineTransform, {x: point[0], y: point[1]})))
+            ret[unit.unitId] = pts
+        }
+        return ret
+    }, [affineTransform, transform, units])
+
+    const paintUnits = useCallback((ctxt: CanvasRenderingContext2D, props: any) => {
+        // const markerRadius2 = markerRadius * radiusScale
+        const markerRadius2 = markerRadius
+        ctxt.clearRect(0, 0, ctxt.canvas.width, ctxt.canvas.height)
+        const drawUnit = (pts: {x: number, y: number}[], color: string) => {
+            ctxt.fillStyle = color
+            ctxt.strokeStyle = 'black'
+            for (let i = 0; i < pts.length; i++) {
+                const {x, y} = pts[i]
+                ctxt.beginPath()
+                ctxt.ellipse(x, y, markerRadius2, markerRadius2, 0, 0, circle)
+                ctxt.fill()
+                ctxt.stroke()
+            }
+        }
+        for (let unit of filteredUnits) {
+            const pts = pixelPointsByUnit[unit.unitId]
+            const col = selectedUnitIds.size === 0 || selectedUnitIds.has(unit.unitId) ? (
+                colorForUnitId(idToNum(unit.unitId))
+            ) : 'rgb(220, 220, 220)'
+            drawUnit(pts, col)
+        }
+    }, [filteredUnits, selectedUnitIds, pixelPointsByUnit])
+
+    const electrodeGeometryCanvas = useMemo(() => {
+        return <BaseCanvas 
+            width={width}
+            height={height}
+            draw={paintElectrodes}
+            drawData={emptyDrawData}
+        />
+    }, [width, height, paintElectrodes])
+
+    const unitsCanvas = useMemo(() => {
+        return <BaseCanvas 
+            width={width}
+            height={height}
+            draw={paintUnits}
+            drawData={emptyDrawData}
+        />
+    }, [width, height, paintUnits])
+
+    const handleSelectRect = useCallback((r: Vec4, {ctrlKey}: {ctrlKey: boolean}) => {
+        const r2 = applyAffineTransformToRectInv(affineTransform, r)
```

Comment:
We should talk through the affine transform logic at some point--I think this is another case where we have code with similar effects in several places. (And again, I'd prefer to do it as explicit matrix multiplication from a library instead of writing our own.)

In general I've found it clearer and more responsive to do most of the work in pixel space, particularly for UI interactions. Here, I'm not sure the effect of applying an inverse affine transform to what I assume is the selection rectangle--particularly as the rectangular area comparison a few lines below is being done with pixels... maybe we don't wind up actually having a transform to apply to the rectangle?

Reply:
okay we can discuss. Generally I like only solving a performance problem once it is identified as a problem.